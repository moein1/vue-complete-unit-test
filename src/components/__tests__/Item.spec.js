import {
  shallowMount,
  RouterLinkStub
} from '@vue/test-utils'
import Item from '../Item.vue'
import merge from 'lodash.merge'

function createWrapper (overrides) {
  const defaultMountingOptions = {
    stubs: {
      RouterLink: RouterLinkStub
    },
    propsData: {
      item: {}
    }
  }
  return shallowMount(Item, merge(defaultMountingOptions, overrides))
}

describe('Item.vue', () => {
  test('renders the hostname', () => {
    const item = {
      url: 'https://some-url.com/with-paths'
    }

    const wrapper = createWrapper({
      propsData: {
        item
      }
    })
    expect(wrapper.text()).toContain('(some-url.com)')
  })

  test('renders item.score', () => {
    const item = {
      score: 10
    }
    const wrapper = createWrapper({
      propsData: {
        item
      }
    })
    expect(wrapper.text()).toContain(item.score)
  })

  test('renders item.by', () => {
    const item = {
      by: 'some author'
    }
    const wrapper = createWrapper({
      propsData: {
        item
      }
    })
    expect(wrapper.text()).toContain(item.by)
  })

  test('renders a link to the item.url with item.title as text', () => {
    const item = {
      url: 'http://some-url.com',
      title: 'some-title'
    }
    const wrapper = createWrapper({
      propsData: {
        item
      }
    })
    const a = wrapper.find('a')
    expect(a.text()).toBe(item.title)
    expect(a.attributes().href).toBe(item.url)
  })

  test('renders the time since the last post', () => {
    const dateNow = jest.spyOn(Date, 'now')
    const dateNowTime = new Date('2018')

    dateNow.mockImplementation(() => dateNowTime)

    const item = {
      time: (dateNowTime / 1000) - 600
    }
    const wrapper = createWrapper({
      propsData: {
        item
      }
    })
    dateNow.mockRestore()
    expect(wrapper.text()).toContain('10 minutes ago')
  })

  test('renders correctly', () => {
    const dateNow = jest.spyOn(Date, 'now')
    const dateNowTime = new Date('2018')

    dateNow.mockImplementation(() => dateNowTime)

    const item = {
      by: 'eddyerburgh',
      id: 11122233,
      score: 10,
      time: (dateNowTime / 1000) - 600,
      title: 'vue-test-utils is released',
      type: 'story',
      url: 'https://vue-test-utils.vuejs.org/'
    }
    const wrapper = createWrapper({
      propsData: {
        item
      }
    })
    dateNow.mockRestore()
    expect(wrapper.element).toMatchSnapshot()
  })

  test('renders correctly when item has no url', () => {
    const dateNow = jest.spyOn(Date, 'now')
    const dateNowTime = new Date('2018')

    dateNow.mockImplementation(() => dateNowTime)

    const item = {
      by: 'eddyerburgh',
      id: 11122233,
      score: 10,
      time: (dateNowTime / 1000) - 600,
      title: 'vue-test-utils is released'
    }
    const wrapper = createWrapper({
      propsData: {
        item
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
