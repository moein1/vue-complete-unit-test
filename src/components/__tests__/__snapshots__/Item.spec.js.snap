// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item.vue renders correctly 1`] = `
<li
  class="news-item"
>
  <span
    class="score"
  >
    10
  </span>
   
  <span
    class="title"
  >
    <a
      href="https://vue-test-utils.vuejs.org/"
      rel="noopener"
      target="_blank"
    >
      vue-test-utils is released
    </a>
     
    <span
      class="host"
    >
       (vue-test-utils.vuejs.org)
    </span>
  </span>
   
  <br />
   
  <span
    class="meta"
  >
    <span
      class="by"
    >
      
      by 
      <a>
        eddyerburgh
      </a>
    </span>
     
    <span
      class="comments-link"
    >
      
      | 
      <a>
         comments
      </a>
    </span>
     
    <span>
      
      10 minutes ago
    
    </span>
  </span>
</li>
`;

exports[`Item.vue renders correctly when item has no url 1`] = `
<li
  class="news-item"
>
  <span
    class="score"
  >
    10
  </span>
   
  <span
    class="title"
  >
    <a
      rel="noopener"
      target="_blank"
    >
      vue-test-utils is released
    </a>
     
    <span
      class="host"
    >
       ()
    </span>
  </span>
   
  <br />
   
  <span
    class="meta"
  >
    <span
      class="by"
    >
      
      by 
      <a>
        eddyerburgh
      </a>
    </span>
     
    <span
      class="comments-link"
    >
      
      | 
      <a>
         comments
      </a>
    </span>
     
    <span>
      
      10 minutes ago
    
    </span>
  </span>
</li>
`;
